src/dataflow/src/render/debezium.rs

```diff
@@ -0,0 +1,665 @@
+                                    },
+                                    Some(Err(err)) => Some(Err(DataflowError::from(err))),
+                                    None => continue,
+                                };
+
+                                let retraction = match value {
+                                    Some(value) => {
+                                        session.give((value.clone(), cap.time().clone(), 1));
+                                        current_values.insert(key, value)
+                                    }
+                                    None => current_values.remove(&key),
+                                };
+
+                                if let Some(res) = retraction {
+                                    session.give((res, cap.time().clone(), -1));
+                                }
+                            }
+                        }
+                        gauge.set(current_values.len() as u64);
+                    }
+                })
+                .as_collection()
+        }
+        _ => input
+            .unary(Pipeline, "envelope-debezium", move |_, _| {
+                let mut state = DebeziumDeduplicationState::new(envelope.clone());
+                let mut data = vec![];
+                move |input, output| {
+                    while let Some((cap, refmut_data)) = input.next() {
+                        let mut session = output.session(&cap);
+                        refmut_data.swap(&mut data);
+                        for result in data.drain(..) {
+                            let key = match result.key.transpose() {
+                                Ok(key) => key,
+                                Err(err) => {
+                                    session.give((Err(err.into()), cap.time().clone(), 1));
+                                    continue;
+                                }
+                            };
+                            let value = match result.value {
+                                Some(Ok(value)) => value,
+                                Some(Err(err)) => {
+                                    session.give((Err(err.into()), cap.time().clone(), 1));
+                                    continue;
+                                }
+                                None => continue,
+                            };
+
+                            let should_use = match state {
+                                Some(ref mut s) => s.should_use_record(
+                                    key,
+                                    &value,
+                                    result.position,
+                                    result.upstream_time_millis,
+                                    &debug_name,
+                                ),
+                                None => true,
+                            };
+
+                            if should_use {
+                                match value.iter().nth(before_idx).unwrap() {
```

Comment:
comment thread from before disappeared.  I didn't notice that the `before` had a `-1` and the `after` had a `+1` so you're right -- my question didn't make sense.  all good!